### PR TITLE
Logging improvements

### PR DESCRIPTION
**Pull request by Alkar778**
### Description
This PR reduces the size of cache.log file

### Manual testing scenarios
1. Create a product
2. Go to var/log/cache.log file
Expected Result
1. File Exists
2. File Contain invalidated tags and compressed debug trace 

### Release notes

Reduced the size of the cache.log file to improve performance.

### Associated documentation updates

Add link to Magento DevDocs PR or Issue, if needed.

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] Pull request introduces user-facing changes and includes meaningful updates for any required release notes and documentation changes
 - [ ] All commits are accompanied by meaningful commit messages

### CODE REVIEW

**Comment by BarnyShergold** [comment]
@magento run all tests

**Review by BarnyShergold — CHANGES_REQUESTED** [review]
Add proper description, notes etc
Ensure Travis tests pass.

**Comment by Alkar778** [comment]
@BarnyShergold 
The description has been updated.
Travis is failing not through my fault

File: Model/Cache/InvalidateLogger.php [hunk 1]
```diff
@@ -13,20 +13,67 @@
...
-                'trace' => debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS)
-            ];
+        if (is_array($invalidateInfo) && isset($invalidateInfo['tags'])) {
+            foreach ($invalidateInfo['tags'] as $tag) {
+                if (in_array(strtolower($tag), $this->tagsToLog)) {
+                    if (function_exists('gzcompress')) {
+                        $invalidateInfo['trace'] =  bin2hex(
+                            gzcompress(
+                                print_r(
+                                    debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS),
```
Comment (kandy) [inline Model/Cache/InvalidateLogger.php]: Can I ask you to remove from Trace the lines contains classes that match patterns "Magento\ObjectManager" and "Abstract\w*$" to decries trace size even more?

File: Model/Cache/InvalidateLogger.php [hunk 2]
```diff
@@ -13,20 +13,67 @@
...
     {
-        if (is_array($invalidateInfo)) {
-            $invalidateInfo['trace'] = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
-        } elseif (is_string($invalidateInfo)) {
-            $invalidateInfo = [
-                'main' => $invalidateInfo,
-                'trace' => debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS)
-            ];
+        if (is_array($invalidateInfo) && isset($invalidateInfo['tags'])) {
+            foreach ($invalidateInfo['tags'] as $tag) {
```
Comment (kandy) [inline Model/Cache/InvalidateLogger.php:88]: Should we get trace only once even if multiple tags is matched?

**Review by kandy — CHANGES_REQUESTED** [review]
Please check  the comment

File: Model/Cache/InvalidateLogger.php (hunk 1 above)
Comment (Alkar778) [reply Model/Cache/InvalidateLogger.php]: Not only the ObjectManager class has been removed from the trace.
Please take a look at a second commit.

File: Model/Cache/InvalidateLogger.php (hunk 2 above)
Comment (Alkar778) [reply Model/Cache/InvalidateLogger.php:88]: Yes, good catch. 
We need to log trace only once.

File: Model/Cache/InvalidateLogger.php (hunk 1 above)
Comment (kandy) [reply Model/Cache/InvalidateLogger.php]: Thanks
Even better than I proposed

**Review by kandy — DISMISSED** [review]

**Review by BarnyShergold — DISMISSED** [review]
Accepted that Travis is failing for unknown reasons

**Comment by oshmyheliuk** [comment]
@Alkar778 I've created separate PR from this pr for test running and static tests failed
```
FILE: /home/travis/build/magento/magento-cloud-components/Model/DebugTrace.php
------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
------------------------------------------------------------------------------
 6 | ERROR | [x] Missing strict_types declaration
------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------
FILE: ...is/build/magento/magento-cloud-components/Model/Cache/InvalidateLogger.php
--------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
--------------------------------------------------------------------------------
 78 | ERROR | Code must not contain multiple empty lines in a row; found 2
    |       | empty lines
--------------------------------------------------------------------------------
```
Please fix.

**Comment by Alkar778** [comment]
@oshmyheliuk 
code style fixed

File: Model/DebugTrace.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+namespace Magento\CloudComponents\Model;
```
Comment (BarnyShergold) [inline Model/DebugTrace.php:6]: Need to add -
```suggestion
declare(strict_types=1);


```

Comment (BarnyShergold) [inline Model/DebugTrace.php:9]: Need to add -

```suggestion
declare(strict_types=1);

namespace Magento\CloudComponents\Model;
```

Comment (oshmyheliuk) [reply Model/DebugTrace.php:9]: Yes, still failing without `declare`
```
FILE: /home/travis/build/magento/magento-cloud-components/Model/DebugTrace.php
------------------------------------------------------------------------------
FOUND 1 ERROR AFFECTING 1 LINE
------------------------------------------------------------------------------
 6 | ERROR | [x] Missing strict_types declaration
------------------------------------------------------------------------------
PHPCBF CAN FIX THE 1 MARKED SNIFF VIOLATIONS AUTOMATICALLY
------------------------------------------------------------------------------
```

**Comment by Alkar778** [comment]
@oshmyheliuk 
I've pushed additional commit. Please review

**Comment by oshmyheliuk** [comment]
@Alkar778 Static tests are ok, but functional failed
```
 [Exception]                                                                                                    
  Notice: Undefined index: file in /app/vendor/magento/magento-cloud-components/Model/DebugTrace.php on line 65 
```

File: Model/DebugTrace.php [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * Returns debug back trace
+     *
+     * @return array
+     */
+    public function getTrace()
+    {
+        $trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+        foreach ($trace as $index => $line) {
```
Comment (oshmyheliuk) [inline Model/DebugTrace.php:62]: Need to add validation if elements of array `$line` is set.
```
$line['function']
$line['class']
$line['file']
```

Comment (Alkar778) [reply Model/DebugTrace.php:62]: It always set 
https://www.php.net/manual/ru/function.debug-backtrace.php

Comment (Alkar778) [reply Model/DebugTrace.php:62]: How can I reproduce case when the **file** key undefined?

Comment (oshmyheliuk) [reply Model/DebugTrace.php:62]: Such error happening during `setup:install`.
https://travis-ci.com/github/magento/magento-cloud-components/jobs/378018983

Comment (Alkar778) [reply Model/DebugTrace.php:62]: Fixed

**Review by BarnyShergold — APPROVED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Review by shiftedreality — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**